Reimagine Sharing: Export Tweaks

bjtitus opened the pull request:
| 📘 Part of: #1910 |
|:---:|

This makes several changes to the final shared asset on Instagram and changes the flow of clip export to load _after_ selecting a service instead of before.

Other changes:
* Increases resolution of final exported image
* Fixes audio in saved video clip
* Fixes audio clip export to Instagram
* Enhances progress reporting to include composition export - this still doesn't include the file export and could use further tweaking.

## To test

### Video Export

> [!NOTE]
> Enable the `newSharing` feature flag to test this

* Navigate to an episode
* Tap the "Share" button
* Tap "Clip"
* Tap "Next"
* Tap the "More" share option
* Verify that loading appears at this point
* Verify that system share sheet opens after completion
* Try saving the video. Verify that the video is saved properly
* * Try sharing to another service. Verify that a link and image file is shared - **NOTE**: Not all services behave the same. Tumblr does a good job of showing all shared media, X will only show the video, for instance.
* Tap "Copy Link" and verify the link is copied to the clipboard

### Audio Export

* Tap "Edit clip" and swipe to the audio style
* Tap the "More" button 
* Verify that the link and audio file are shared.
* Tap the "Stories" integration and verify that a video file with an audio symbol is shared to Instagram
* Tap "Copy Link" and verify the link is copied to the clipboard

### Image Export

* Navigate to an episode
* Tap the "Share" button
* Tap "Episode" or "podcast
* Tap "Next"
* Tap the "More" share option
* Verify that system share sheet opens after completion with a link and image asset
* Try saving the image. Verify that the image is saved properly
* Try sharing to another service. Verify that a link and image file is shared - **NOTE**: Not all services behave the same. Tumblr does a good job of showing all shared media but Messages will only show the link, and X will only show the image, for instance.
* Tap "Copy Link" and verify the link is copied to the clipboard

## Checklist

- [x] I have considered if this change warrants user-facing release notes and have added them to `CHANGELOG.md` if necessary.
- [x] I have considered adding unit tests for my changes.
- [x] I have updated (or requested that someone edit) [the spreadsheet](https://docs.google.com/spreadsheets/d/107jqrutZhU0fVZJ19SBqxxVKbV2NWSdQC9MFYdLiAxc/edit?usp=sharing) to reflect any new or changed analytics.

bjtitus committed:
Remove separate crop export & use background color

bjtitus committed:
Change sharing export depending on destination

bjtitus committed:
Change share image sizes

bjtitus committed:
Improve exported image resolution

bjtitus committed:
Tweak blob sizing and effects for less banding

bjtitus committed:
Improve progress reporting

bjtitus committed:
AVAssetWriterInput error handler + continue after finish

* The AVAssetWriterInput should be marked as finished when an error occurs
* The Continuation should only be called when finished writing, not after each block call. This prevents a crash where continuation could be resumed multiple times.

bjtitus committed:
Export video for audio instagram share

bjtitus committed:
Improve shared image resolution

dangermattic commented:
<!--
  0 Errors
  2 Warnings: View files have been modified,...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          2 Warnings
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">View files have been modified, but no screenshot or video is included in the pull request. Consider adding some for clarity.</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is assigned to the milestone <a href="https://github.com/Automattic/pocket-casts-ios/milestone/58">7.72</a>. The due date for this milestone has already passed.<br />
Please assign it to a milestone with a later deadline or check whether the release for this milestone has already been finished.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

bjtitus committed:
Change "Clip" button to "Next"

danielebogo reviewed (changes requested):
Hey Brandon,
I have to finish the code review but I had a crash while testing it the Audio Export for Instagram stories (crash below). The audio saving and sharing trough Whatsapp worked though.

<img width="1912" alt="Screenshot 2024-09-02 at 14 57 43" src="https://github.com/user-attachments/assets/15668720-62e2-47ac-93f2-31160bf78dd6">

```
Fraction completed: 0.002
Fraction completed: 0.004
Fraction completed: 0.006
Fraction completed: 0.008
Fraction completed: 0.01
Fraction completed: 0.012
Fraction completed: 0.014
Fraction completed: 0.016
Fraction completed: 0.018
Fraction completed: 0.02
Fraction completed: 0.022
Fraction completed: 0.024
Fraction completed: 0.026
Fraction completed: 0.028
Fraction completed: 0.03
Fraction completed: 0.032
Fraction completed: 0.034
Fraction completed: 0.036
Fraction completed: 0.038
Fraction completed: 0.04
Fraction completed: 0.042
Fraction completed: 0.044
Fraction completed: 0.046
Fraction completed: 0.048
Fraction completed: 0.05
Fraction completed: 0.052
*** Terminating app due to uncaught exception 'NSInternalInconsistencyException', reason: '*** -[AVAssetWriterInputPixelBufferAdaptor appendPixelBuffer:withPresentationTime:] A pixel buffer cannot be appended when readyForMoreMediaData is NO.'
*** First throw call stack:
(0x19d284f20 0x19510b2b8 0x1aec00970 0x1aebfe480 0x1aec03754 0x105b40c08 0x105b44115 0x105b42909 0x105b4447d 0x10500b579 0x10500b859 0x104d70469 0x104d705c5 0x1a85d2775)
libc++abi: terminating due to uncaught exception of type NSException
```

The Video and Image export worked as expected!

bjtitus committed:
Check isReadyForMoreMediaData after buffer creation

Because the buffer creation is delayed, the isReadyForMoreMediaData might change status in the time. We should check again just before appending

bjtitus committed:
Replace console with file log for media export

bjtitus commented:
@danielebogo I'm having trouble reproducing on my test devices. I suspect this could be `readyForMoreMediaData` becoming false in the time it takes to create the pixel buffer. Could you retest with the latest changes?

danielebogo commented on podcasts/Sharing/SharingFooterView.swift:
```diff
@@ -64,8 +69,16 @@ struct SharingFooterView: View {
+                do {
+                    try await destination.share(option, style: style, clipTime: clipTime, clipUUID: clipUUID, progress: $progress, source: source)
+                } catch let error {
+                    guard Task.isCancelled == false  else { return }
```
Nit: can this one be `if Task.isCancelled { return }`?

danielebogo commented on podcasts/Sharing/SharingFooterView.swift:
```diff
@@ -64,8 +69,16 @@ struct SharingFooterView: View {
+            isExporting = true
+            shareTask = Task.detached {
+                do {
+                    try await destination.share(option, style: style, clipTime: clipTime, clipUUID: clipUUID, progress: $progress, source: source)
```
I get 2 warnings here:
```
Non-sendable type 'Binding<Float?>' in implicitly asynchronous access to main actor-isolated property '$progress' cannot cross actor boundary
```
```
Passing argument of non-sendable type 'Binding<Float?>' into main actor-isolated context may introduce data races
```

bjtitus replied:
Marking this as `MainActor` seemed to fix this.

danielebogo commented on podcasts/Sharing/SharingModal.swift:
```diff
@@ -168,39 +177,47 @@ extension SharingModal.Option {
-                fileURL = progress.croppedURL
+    func mediaData(imageInfo: ShareImageInfo, style: ShareImageStyle, episode: Episode, clipTime: ClipTime, destination: ShareDestination, progress: Binding<Float?>) async throws -> Any? {
+        let nsProgress = Progress(totalUnitCount: 100)
+        let observation = nsProgress.observe(\.fractionCompleted) { [progress] inProgress, change in
+            Task.detached { @MainActor in
+                guard Task.isCancelled == false && inProgress.isCancelled == false else { return }
+                FileLog.shared.addMessage("Media Exporter: Fraction completed: \(inProgress.fractionCompleted)")
+                progress.wrappedValue = Float(inProgress.fractionCompleted)
```
Few warnings here:
`let observation` seems not be used.
Also
```
Capture of 'progress' with non-sendable type 'Binding<Float?>' in a `@Sendable` closure
```

danielebogo commented on podcasts/Sharing/AnimatedShareImageView.swift:
Leaving a comment here but in
```
class AnimationProgress: ObservableObject {
    @Published var progress: Double = 0 // O-1
}
```
seems to be updated from a bg thread: I get this `Publishing changes from background threads is not allowed; make sure to publish values from the main thread (via operators like receive(on:)) on model updates.`

bjtitus replied:
Marked this as `@MainActor` so it shouldn't do this, I think.

danielebogo reviewed (changes requested):
The crash seems gone now! 
Just left some comments. Other than that I think we can go ahead and merge it

bjtitus committed:
Fix unused variable warning

This observation property needs to remain around for the lifetime of this function. I could use withExtendedLifetime but it doesn't have an async version so I think invalidating makes sense.

bjtitus committed:
Mark AnimationProgress as MainActor

bjtitus committed:
PR feedback

bjtitus committed:
Fix Sendable errors for progress

danielebogo reviewed (approved):
All good! thanks for addressing those feedback! :shipit:
